lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -450,28 +489,121 @@ assignMemoryLayouts(llvm::SmallVector<std::tuple<Operation *, int, Operation *>>
+    if (!seen.insert(op).second)
+      return;
+    if (op != baseOp)
+      users.push_back(op);
+    if (isa<tt::LoadOp, tt::ExperimentalDescriptorLoadOp>(op)) {
+      if (op != baseOp)
+        // Stop recursion when hitting a LoadOp.
+        return;
+    }
+    for (Operation *user : op->getUsers())
+      if (user->getBlock() == op->getBlock())
+        dfs(user, baseOp);
+  };
+  dfs(baseOp, baseOp);
+  return users;
+}
+
+static llvm::MapVector<Operation *, LoadInfo>
+assignMemoryLayouts(scf::ForOp &forOp, tt::CoarseSchedule &schedule,
+                    tt::ModuleAxisInfoAnalysis &axisInfoAnalysis) {
+  llvm::MapVector<Operation *, LoadInfo> loadToInfo;
+
+  // Go through all loads in the loop, check to see if they are pipelined.
+  llvm::DenseSet<Operation *> loadsToPipeline;
+  for (auto &op : forOp.getBody()->without_terminator()) {
+    if (!isa<tt::LoadOp>(op) && !isa<tt::ExperimentalDescriptorLoadOp>(op))
+      continue;
+    if (loadToInfo.count(&op))
+      // TODO pawel: err, we'd need to verify that the distance is the same
+      continue;
+    if (!schedule.count(&op))
       continue;
+
+    // Check stage for uses. If any use is in a different stage, treat it
+    // as a pipelined load.
+    auto users = getTransitiveUserInBlock(&op);
+    bool isPipelined = false;
+    auto [sLoad, _cLoad] = schedule[&op];
+    for (auto user : users) {
+      if (!schedule.count(user))
+        continue;
+      auto [stage, _cluster] = schedule[user];
+      if (stage != sLoad) {
+        isPipelined = true;
+        break;
+      }
     }
-    loadToInfo[op] = loadInfo;
+    if (!isPipelined)
+      continue;
+
+    loadsToPipeline.insert(&op);
+    LoadInfo loadInfo;
+    for (auto use : users) {
+      if (use->hasTrait<OpTrait::DotLike>()) {
+        loadInfo.usedByDot = true;
+        if (loadIsMMAv3(&op)) {
+          loadInfo.loadIsMMAV3 = true;
+          loadInfo.sharedEncoding =
+              getSharedEncoding(&op, /*loadIsMMAv3=*/true).value_or(nullptr);
+        } else if (isa<tt::ExperimentalDescriptorLoadOp>(op)) {
+          loadInfo.sharedEncoding =
+              getSharedEncoding(&op, /*loadIsMMAv3=*/true).value_or(nullptr);
+        } else if (auto dot = dyn_cast<tt::DotOp>(use)) {
+          loadInfo.sharedEncoding =
+              getSharedEncIfAllUsersAreDotEnc(op.getResult(0))
+                  .value_or(nullptr);
+        }
+      }
+
+      // If we still don't have a shared encoding, try a "generic" shared
+      // encoding.
+      if (!loadInfo.sharedEncoding && !isa<ttng::WarpGroupDotOp>(use)) {
+        loadInfo.sharedEncoding =
+            getSharedEncoding(&op, /*isMMAV3=*/loadInfo.loadIsMMAV3)
+                .value_or(nullptr);
+        if (auto loadOp = dyn_cast<tt::LoadOp>(op))
+          loadInfo.blockedEncoding =
+              getBlockedEncoding(loadOp, axisInfoAnalysis);
+      }
+
+      // If that still didn't work, bail on pipelining this load.
+      if (!loadInfo.sharedEncoding)
```

Comment:
That should never happen, right?

Reply:
I think so. I currently assert outside of the loop to make sure all pipelined loads have shared encoding. I was worried that maybe one use has a sharedEncoding and another use doesn't.

Reply:
Tried adding an assert here, it passed the lit tests.
assert(loadInfo.sharedEncoding);
Should we go ahead adding the assert here?

Reply:
I wonder if this wouldn't trigger in cases where load is used by some arithmetic instruction transitively on the way to dot?

Reply:
Yeah that will trigger the assert. So leaving it like this in the patch will be good?

Reply:
Yes, I think its good as it is.